src/main/java/io/debezium/connector/jdbc/ReducedRecordBuffer.java

```diff
@@ -0,0 +1,77 @@
+/*
+ * Copyright Debezium Authors.
+ *
+ * Licensed under the Apache Software License version 2.0, available at http://www.apache.org/licenses/LICENSE-2.0
+ */
+package io.debezium.connector.jdbc;
+
+import org.apache.kafka.connect.data.Schema;
+import org.apache.kafka.connect.data.Struct;
+import org.apache.kafka.connect.errors.ConnectException;
+
+import java.util.ArrayList;
+import java.util.Map;
+import java.util.List;
+import java.util.Objects;
+import java.util.HashMap;
+
+/**
+ * A reduced implementation buffer of {@link SinkRecordDescriptor}.
+ * It reduces events in buffer before submit to external database.
+ *
+ * @author Gaurav Miglani
+ */
+public class ReducedRecordBuffer implements Buffer {
+
+    private final JdbcSinkConnectorConfig connectorConfig;
+    private Schema keySchema;
+    private Schema valueSchema;
+
+    private final Map<Struct, SinkRecordDescriptor> records = new HashMap<>();
+
+    public ReducedRecordBuffer(JdbcSinkConnectorConfig connectorConfig) {
+        this.connectorConfig = connectorConfig;
+    }
+
+    @Override
+    public List<SinkRecordDescriptor> add(SinkRecordDescriptor recordDescriptor) {
+        ArrayList<SinkRecordDescriptor> flushed = new ArrayList<>();
+
+        if (records.isEmpty()) {
+            keySchema = recordDescriptor.getKeySchema();
+            valueSchema = recordDescriptor.getValueSchema();
+        }
+
+        if (!Objects.equals(keySchema, recordDescriptor.getKeySchema()) || !Objects.equals(valueSchema, recordDescriptor.getValueSchema())) {
+            keySchema = recordDescriptor.getKeySchema();
+            valueSchema = recordDescriptor.getValueSchema();
+            flushed.addAll(flush());
+        }
+
+        Struct keyStruct = recordDescriptor.getKeyStruct(connectorConfig.getPrimaryKeyMode());
+        if (keyStruct != null) {
+            records.put(keyStruct, recordDescriptor);
+        }
+        else {
+            throw new ConnectException("No struct-based primary key defined for record key/value, reduction buffer require struct based primary key");
```

Comment:
I am not fully convinced on this. What will happen if the source table has no primary defined? Maybe a separate test case would be good for it.

Reply:
ok working on a test case where there is no primary key, but output is correct? ideally ReducedRecordBuffer should fail in case no primary key, correct @jpechane ?

Reply:
Hi @jpechane , the method `getKeyStruct` return null in case no primary key is defined, so i think my last test case cover that.

Reply:
No, it should not fail, it is IMHO valid case and just buffering should not be done.

Reply:
I'm little confused here, or may be i'm missing something here, need more  help to fix, AFAIK in the code flow, first we receive kafka connect `SinkRecord` that can be from multiple kafka topics assuming (topic1 with pkey and topic2 with no pkey), and let's say we are using RECORD_KEY mode, first we build SinkRecordDescriptor and apply flushing logic. Now if user is using `use.reduction.buffer` = true, ReducedRecordBuffer works on a Map, now i'm not sure what map key should be in case of topic2  records, so initially I raised  an exception. Can you please help more, I will make changes.

Reply:
@gaurav7261 In the case of `topic2` record there will be no reduction, it will just pass-through like the size of batch be 1 or `use.reduction.buffer` not set.

Reply:
so should I have have one more instance variable apart from map, like arraylist for this, and in flush method, club them record.values + list  in ReducedRecordBuffer implementation ?

Reply:
@jpechane waiting for your review here as well

Reply:
I think second option `use.reduction.buffer` not set is correct for this, thanks , making changes

Reply:
changes done, pushing, also i have same double check in ReducedRecordBuffer for same, let me know if that is needed, otherwise i can remove that, no issue